### PR TITLE
`video` kirbytag supports local and remote videos #3104

### LINKED ISSUE #3104: Video kirbytag supporting local and remote video
Proof of concept plugin: https://github.com/getkirby/getkirby.com/pull/1188

### PR DESCRIPTION
**Pull request by afbora**
## Describe the PR

This PR implements the https://github.com/getkirby/getkirby.com/pull/1188 PR with some improvements, bug fixes and unit tests.

## 🎉 Features

Supported local and remote video files by extending video tag that supporting only youtube and vimeo videos. Youtube and vimeo features have not been changed.

### New Attributes:
- autoplay
- controls (default: true)
- loop
- muted *
- poster
- preload (auto, metadata, none)

_* muted will enabled if autoplay enabled and muted is not defined_

Usages:

````markdown
# local
(video: local-video.mp4)

# remote
(video: https://www.getkirby.com/sample-video.mp4)

# example 1
(video: local-video.mp4  autoplay: true)

# example 2
(video: local-video.mp4 controls: false autoplay: true loop: true)

# example 3
(video: local-video.mp4 poster: cover.jpg)

# example 4
(video: local-video.mp4 preload: auto)

# example 5
(video: https://www.getkirby.com/sample-video.mp4 muted: true controls: false autoplay: true)

# example 6
(video: local-video.mp4 poster: https://www.getkirby.com/sample-cover.jpg)

````

## Related issues

- Fixes #3104 

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Added in-code documentation (if needed)
- [x] CI passes (runs automatically when the PR is created or run `composer ci` locally)
  Running locally requires PHPUnit, PHP-CS-Fixer, Psalm, PHPCPD and PHPMD.

- [x] Checked whether the PR needs documentation, if needed added to the [release docs checklist](https://github.com/getkirby/getkirby.com/pulls)

### CODE REVIEW

File: src/Cms/File.php [hunk 1]
```diff
@@ -248,12 +248,16 @@ public function dragText(string $type = null, bool $absolute = false): string
         if ($type === 'markdown') {
             if ($this->type() === 'image') {
                 return '![' . $this->alt() . '](' . $url . ')';
+            } elseif ($this->type() === 'video') {
+                return '[![ALT TEXT](IMAGE URL)](VIDEO URL)';
```
Comment (afbora) [inline src/Cms/File.php]: Any idea about that?

Comment (bastianallgeier) [reply src/Cms/File.php]: Hm, this is tricky. There's no video tag in markdown. We could only inject an html video tag instead.

Comment (afbora) [reply src/Cms/File.php]: @bastianallgeier You mean like following. 
So actually rendered `video` tag in directly panel text editor?

**Write output to editor;**

If local or remote video

    <figure class="video"><video controls><source src="LOCAL-REMOTE-VIDEO-URL" type="video/mp4"></video></figure>

If youtube or vimeo video
    
    <figure class="video"><iframe allowfullscreen src="YOUTUBE-VIMEO-URL"></iframe></figure>

Comment (afbora) [reply src/Cms/File.php]: Or should we just link to the file (local video) or video url (youtube, vimeo, remote video)?

Comment (bastianallgeier) [reply src/Cms/File.php]: It's definitely not ideal. Another option could be to add a simple markdown link to the video instead. I don't know what makes more sense from a user perspective. All HTML is quite annoying in the context of markdown, but the link is just a very bad compromise either. Maybe keep it simple for this first implementation and only render the markdown link. Then we wait for feedback.

Comment (distantnative) [reply src/Cms/File.php]: I would go for the link. There is no video tag in Markdown that's the reality. Putting a lot of HTML in it should be a conscious decision by the user

**Comment by afbora** [comment]
Now ready to review 🚀

File: config/tags.php [hunk 2]
```diff
@@ -236,23 +239,117 @@
...
+                    (isset($tag->{$key}) === false || $tag->{$key} === null)
+                ) {
+                    $tag->{$key} = $value;
+                }
+            }
+
+            // generates tag attributes helper inline function
+            // for ex: `autoplay` will not work if `false` is a `string` instead of a `boolean`
+            // this is why the type converter is used in function
+            $generateTagAttributes = function ($tag) {
```
Comment (distantnative) [inline config/tags.php]: Do we need an inline function here if we only call it once? Could be just the code, without putting it in a function, or?

Comment (afbora) [reply config/tags.php]: You are absolutely right. Previously, I was calling this inline function from two places. So I needed an inline function. After improvement I used it in one place but forgot it there 🙈

Comment (afbora) [reply config/tags.php]: Done

**Review by bastianallgeier — APPROVED** [review]
I love this! I'm sure this is going to make a lot of users happy.

**Pull request merged**